### ISSUE TITLE
Change Global $topic Variable Name or Maybe create Topic Class

### ISSUE THREAD

**Issue by eSilverStrike** [Geeklog-Core/geeklog#872]
Right now to keep track of the current user topic we use the global variable $topic.

The problem is that this is a popular variable name which can make it confusing. For example Polls and Links use this variable name for their own internal categories / topic.

We should probably change the global variable name (not url variables) to something like $gl-topic-id to make sure it is easy to spot or **better yet create a topic class to set and get the current topic id and any other functionality**.

This should not affect third party plugins as I don't know of any that use the global topic variable (only core ones do). 

For setting $topic variable  Plugins should be using the function:

TOPIC_getTopic

but to retrieve the topic they just access the global $topic variable so there might be an issue here....

**Comment by mystralkk** [comment]
I guess the Session class will handle this issue.  To set the topic, use Session::setVar('topic', topicID), to get the topic, just use Session::getVar('topic', optionalTopicID).

**Comment by eSilverStrike** [comment]
Sounds like a good idea especially since we already have a topic session variable.

**Comment by eSilverStrike** [comment]
I forgot about this session topic variable. I will have to update the forum with the changes I did to use the session variable. @mystralkk has any of Geeklog and the core plugins (that use it) been switch over to use the proposed topic session variable? I did a quick search of the code and it didn't look like it.

**Comment by mystralkk** [comment]
No, I haven't changed any of them.  I once tried to change them, but gave up since there too many of them.

**Comment by eSilverStrike** [comment]
I plan on taking a look into this next then. 

I probably will depreciate the use of the Global topic variable but still keep the functionality there for a while since it is not easy to notify developers of it's depreciation like it is with functions.

I will then use your session class to store the topic variable and update core and the plugins to access it. I will have to be careful going through the code as there is a lot of  local $topic variables that I don't want to touch.

**Comment by eSilverStrike** [comment]
Okay. I took a look at this further and I understand why I was suggesting a topic class (glad I had code comments to take a look at) but it is for sure not the right way to go. Hopefully I can explain this clearly :-)

The topic stored in the session is to allow Geeklog to determine what the topic Id was for the previous page. This is set when the final page is created with COM_createHTMLDocument. I assume it was set in COM_createHTMLDocument to allow for any changes to the $topic variable. The topic session variable is used as a fall back for TOPIC_getTopic if it can't figure out the topic_id when called.

$topic was originally set in lib-common but this had to be removed and set by what ever file was being visited by the user since with url_rewrite enabled, this created issues plus not all plugins are topic aware  (for example the polls plugin). lib-common now only sets a topic default (for all topics) since plugins that do not use topics defaults to this. 

Nothing should really be setting the $topic variable except lib-common at the beginning  to all topics and then TOPIC_getTopic. The problem is TOPIC_getTopic is not called by all pages since not all items are assigned to a topic. We need the topic session variable but we can't use it to store the current topic since all items are not topic aware and we don't know when it was last set (unless we add a flag). 

So the options is we add another topic session variable (or use a session flag) or store the current global $topic variable somewhere else. Since we don't need to remember this from page to page I am thinking of properly formatting the global variable name, something like $_TOPIC_CURRENT (we already have a $TOPICS array) or maybe using the $_USER global array to store it since it deals with the user. Something simple like $_USER['topic_id'] (or $_USER['current_topic_id'])

@mystralkk I figure I will use $_USER['topic_id']. Sound good?

**Comment by eSilverStrike** [comment]
Okay I have converted Geeklog and plugins to use $_USER['topic_id'] instead of global $topic. I want to test this a bit more as their was a lot of little changes so I will probably commit it tomorrow.

I created a function called 

```
// Some plugin may still be using Global $topic variable so check and update $_USER array as needed
_depreciatedCheckGlobalTopicVariableUsed();
```

That checks to see if the Global $topic variable has been set. If it has it will update the $_USER variable. This is used in

lib-common - for COM_createHTMLDocument
lib-comment
index.php - for topics

I hope this will still allow old plugins to use the $topic variable and keep everything functioning until we remove $topic in Geeklog v3.0.0

One other thing before I forget . $_USER['topic_id'] gets updated in 5 spots in Geeklog. 

index.php - When displaying topics 
COM_createHTMLDocument
TOPIC_getTopic
InitSubmission - In article.class.php when setting up the article submission by a user
directory.php - If directory drop down is changed.

**Comment by mystralkk** [comment]
> Okay. I took a look at this further and I understand why I was suggesting a topic class (glad I had code comments to take a look at) but it is for sure not the right way to go. Hopefully I can explain this clearly :-)
> 
> The topic stored in the session is to allow Geeklog to determine what the topic Id was for the previous page. This is set when the final page is created with COM_createHTMLDocument. I assume it was set in COM_createHTMLDocument to allow for any changes to the $topic variable. The topic session variable is used as a fall back for TOPIC_getTopic if it can't figure out the topic_id when called.
> 
> $topic was originally set in lib-common but this had to be removed and set by what ever file was being visited by the user since with url_rewrite enabled, this created issues plus not all plugins are topic aware (for example the polls plugin). lib-common now only sets a topic default (for all topics) since plugins that do not use topics defaults to this.
> 
> Nothing should really be setting the $topic variable except lib-common at the beginning to all topics and then TOPIC_getTopic. The problem is TOPIC_getTopic is not called by all pages since not all items are assigned to a topic. We need the topic session variable but we can't use it to store the current topic since all items are not topic aware and we don't know when it was last set (unless we add a flag).
> 
> So the options is we add another topic session variable (or use a session flag) or store the current global $topic variable somewhere else. Since we don't need to remember this from page to page I am thinking of properly formatting the global variable name, something like $_TOPIC_CURRENT (we already have a $TOPICS array) or maybe using the $_USER global array to store it since it deals with the user. Something simple like $_USER['topic_id'] (or $_USER['current_topic_id'])
> 
> @mystralkk I figure I will use $_USER['topic_id']. Sound good?

I am not sure if I can fully understand about $topic variable.  But using $_USER['topic_id'] sounds to me.

**Comment by eSilverStrike** [comment]
Okay this was a real pain to test and took several fixes before I think I got things right . A lot of small things have been changed. 

All of Geeklog Core and Core Plugins have been updated to use the new way to access the current topic.

So now these 3 functions should be used for the current topic. No more accessing the global variable directly (except by the topic library)

-TOPIC_setTopic should be used by plugins and most other things to set the current topic (now stored in $_USER['topic_id'] variable). Some checks are done here to ensure proper topic
- TOPIC_currentTopic returns the current topic id
- TOPIC_getTopic figures out the current topic for a plugin based on what plugin item has been accessed (since this is stored in the topic assignment table), and the last topic, etc...

I see why it should all be a class but that is to much work at this point and would create huge headaches

The $topic global variable should still work for plugins that have yet to be updated as the topic functions will still set the variable as well (until Geeklog v3.0.0).

@mystralkk more testing of this is welcomed :-)
